content/learning-paths/cross-platform/vectorization-friendly-data-layout/a-more-complex-problem-revisited.md

```diff
@@ -0,0 +1,202 @@
+---
+title: A more complex problem, revisited
+weight: 6
+
+### FIXED, DO NOT MODIFY
+layout: learningpathall
+---
+
+You've seen it mentioned already many times, that changing the data layout can improve performance.
+
+But how you would actually perform such a change and why would it matter?
+
+Consider this modified version of the program:
+
+```C
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <math.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include <arm_neon.h>
+
+#define N           100000 // number of particles
+#define SECONDS     100    // duration in seconds
+#define STEPSPERSEC 1000   // steps per second or time resolution
+
+// The object struct
+typedef struct object {
+  uint32_t id;
+  float weight;
+  void *model_data;
+} object_t;
+
+struct object_list {
+  object_t o[N];
+  float x[N];
+  float y[N];
+  float z[N];
+  float vx[N];
+  float vy[N];
+  float vz[N];
+};
+
+// Helper function to return a random float number from 0.0f to a.
+float randf(float a) {
+  return ((float)rand()/(float)(RAND_MAX)) * a;
+}
+
+const float box_hi[4] = {  10.0f,  10.0f,  10.0f,  10.0f };
+
+void init_objects(struct object_list *objects) {
+  // Give initial speeds and positions
+  for (size_t i=0; i < N; i++) {
+    objects->o[i].id = i;
+    objects->o[i].weight = randf(2.0f);
+    // initial positions in box [-10, 10], velocity in [-1, 1]
+    objects->x[i] = randf(2.0f * box_hi[0]) - box_hi[0];
+    objects->y[i] = randf(2.0f * box_hi[1]) - box_hi[1];
+    objects->z[i] = randf(2.0f * box_hi[2]) - box_hi[2];
+    objects->vx[i] = randf(2.0) - 1.0f;
+    //printf("vx[%ld] = %f\n", i, objects->vx[i]);
+    objects->vy[i] = randf(2.0) - 1.0f;
+    objects->vz[i] = randf(2.0) - 1.0f;
+  }
+}
+
+void simulate_objects(struct object_list *objects, float duration, float step) {
+
+  float current_time = 0.0f;
+  //size_t iterations = 0;
+
+  float32x4_t s = vdupq_n_f32(step);
+  uint32x4_t one = vdupq_n_u32(1);
+  // Collision counters are now per axis
+  uint32x4_t ctr_x = vdupq_n_u32(0);
+  uint32x4_t ctr_y = vdupq_n_u32(0);
+  uint32x4_t ctr_z = vdupq_n_u32(0);
+
+  float32x4_t box_hi_v = vld1q_f32(box_hi);
+  float32x4_t box_lo_v = vnegq_f32(box_hi_v);
+
+  while (current_time < duration) {
+    // Move 4 objects in each iteration
+    for (size_t i=0; i < N; i+= 4) {
+      float32x4_t x_v, y_v, z_v, vx_v, vy_v, vz_v;
+
+      // Load the x,y,z and vx, vy, vz coords for 4 objects
+      x_v = vld1q_f32(&objects->x[i]);
+      y_v = vld1q_f32(&objects->y[i]);
+      z_v = vld1q_f32(&objects->z[i]);
+      vx_v = vld1q_f32(&objects->vx[i]);
+      vy_v = vld1q_f32(&objects->vy[i]);
+      vz_v = vld1q_f32(&objects->vz[i]);
+
+      // Move the 4 objects in each axis
+	  x_v = vmlaq_f32(x_v, vx_v, s);
+      y_v = vmlaq_f32(y_v, vy_v, s);
```

Comment:
some of the performance differences you see here and elsewhere between `clang`  and `gcc` is  because ACLE does not guarantee that `mlaq_f32` is fused.  This intrinsics is provided for backwards compatibility with AArch32 where it was unfused.  Clang fuses it, GCC retains the unfused behavior.

The correct intrinsic to use is `vfmaq_f32` which will fuse.  Also is is possible to get an SVE version here? I think this is a case where SVE could do better. for instance there's no need to do the `AND 1` since a conditional add would do the work. etc.

Reply:
In fact this difference between fused and unfused is commonly misunderstood and may be worth pointing out.

Reply:
Noted, I'll also add SVE (will SVE2 make a difference here?).
It's also good that you mentioned the difference between vmlaq_f32 and vfmaq_f32. I think this warrants an extra section to make the distinction.

Reply:
> will SVE2 make a difference here?
unlikely in this case, but I think we prefer to guide people to SVE2 as a more complete solution in general.

Reply:
I added SVE implementation of the same algorithm, and added the notes for vfma/vmla.